### ISSUE TITLE
Alacritty msg create-window is not focused

### ISSUE THREAD

**Issue by braun-steven** [koekeishiya/yabai#1858]
When using Alacritty and having it already run in the background, creating a new alacritty window with `alacritty msg create-window` results in the new window not being focused (except for when the old focus was already on an alacritty window). You can see this behavior in the following recording:

https://github.com/koekeishiya/yabai/assets/17625341/8f4319a6-812a-4732-ba85-5e97c67c2f81

**Comment by colehpage** [comment]
> When using Alacritty and having it already run in the background, creating a new alacritty window with `alacritty msg create-window` results in the new window not being focused (except for when the old focus was already on an alacritty window). You can see this behavior in the following recording:
> 

@braun-steven - in the same boat - did you end up finding a solution to this? tried some scripting around attempting to grab the most recent window but to no avail....

**Comment by braun-steven** [comment]
The following in skhd did resolve the issue most of the time but I didn't bother too much and went back to iterm2 at some point:

```
# Open alacritty window:
# 1. try to use an exisiting alacritty server to create a new window (fastest)
# 2. if that fails, open the alacritty app
# 3. sleep shortly, such that yabai is able to detect the change (new window)
# 4. focus on latest alacritty window (window with highest id). This fixes an issue where if the focus was previously not on an alacritty window (e.g. safari), the new window would open but the focus stays on the old window (e.g. safari)
# cmd - return : alacritty msg create-window || open -na /Applications/Alacritty.app && sleep 0.5 && yabai -m window --focus $(yabai -m query --windows | jq '[.[] | select(.app == "Alacritty")] | max_by(.id) | .id')
```

**Comment by colehpage** [comment]
Thanks for the response @braun-steven! That's exactly the pathing I was going with last night, minus the sleep which I had a feeling I needed to throw in. I'll give it a try, thanks mate! Been enjoying alacritty for the past year or so and would rather not switch at the moment, but i get it. Have a great rest of your Sunday.

**Comment by colehpage** [comment]
works like a charm! For anyone reading and using `nushell` you can throw the whole pathing into a script and call it a la `lalt - return : bash ~/.config/skhd/open_alacritty.sh`. Otherwise just throw it inline as suggested.

```
#!/usr/bin/env bash

focus_window () {
    WINDOW_ID=$(yabai -m query --windows | jq '[.[] | select(.app == "Alacritty")] | max_by(.id) | .id')
    yabai -m window --focus "${WINDOW_ID}"
}

if ! pgrep -f "alacritty" > /dev/null 2>&1; then
    open -na /Applications/Alacritty.app
else
    alacritty msg create-window && sleep 0.1 && focus_window
fi
```

I've found that a 500ms sleep gives a bit too much of a delay to feel smooth and anything less than 100ms can occasionally miss and focus the last existing window - but at least for now on my machine 100-200ms gives almost completely consistent operation, though that may need to be adjusted for a safer range depending on machine and load...

Anyway - thanks @braun-steven - classic simple sleep was really all that was needed to be tweaked 🙏

**Comment by rcorre** [comment]
I'm having the same problem with `wezterm` after updating to OSX 14. I added the following based on @braun-steven's hack:

```
cmd - return : wezterm start --always-new-process & sleep 0.2 && yabai -m window --focus $(yabai -m query --windows | jq '[.[] | select(.app == "wezterm-gui")] | max_by(.id) | .id')
```

For some reason I needed `&`, not `&&` -- with the latter, it would not focus the new windows.
I also found that a sleep of `0.1` consistently selected the wrong window, often causing a disorienting workspace shift.
This still feels pretty fragile, as @colehpage I imagine it will periodically fail under load, but I also found `0.5` to be noticeably laggy.

**Comment by colehpage** [comment]
@rcorre yeah 100% on the fragility here, it's going to be super system/load dependent. I should have updated here, but I ended up switching to 0.2 and for the past week or so and things have been ~95% accurate focus wise. Testing with some higher sleeps might get you closer to 100% but the lag gets pretty noticeable quickly and I'm very okay with a very occasional prev focus slip to get things snappy the rest of the time. Good shout though for others 👍

**Comment by koekeishiya** [comment]
Use a signal to avoid delay/lag/timing issues..

```
cmd - return : yabai -m signal --add event="window_created" label="alacritty-new-window" app="^Alacritty$" action="yabai -m window \$YABAI_WINDOW_ID --focus && yabai -m signal --remove alacritty-new-window" && (/Applications/Alacritty.app/Contents/MacOS/alacritty msg create-window || open -na /Applications/Alacritty.app)

```

**Comment by colehpage** [comment]
Ahh @koekeishiya the man himself, thanks for chiming in mate. Looks straightforward but I'll give it a try when I get home, appreciate the input.

**Comment by rcorre** [comment]
Thanks @koekeishiya, that's an improvment.
Does closing this mean `yabai` can't/won't handle this?
The workaround is certainly better but still feels somewhat clunky, and doesn't handle if I e.g. launch a GUI app from the terminal.

**Comment by koekeishiya** [comment]
I think this is a newly introduced macOS bug. yabai has never initiated a change in application/window focus without user-input.

Personally I have not experienced this issue with software that I use (although I managed to reproduce this Alacritty issue).

I don't believe this is something that should be fixed through a yabai code change.

The next version of Sonoma should be released in a couple of days; let's see if the issue has been resolved there, and if not I might look into a more general fix.

**Comment by koekeishiya** [comment]
Updated command in case you are "spamming" the hotkey to create multiple terminal windows; the last one should be the final focus in this solution:

```
cmd - return : (LABEL=$(uuidgen) && yabai -m signal --add event="window_created" label="$LABEL" app="^Alacritty$" action="yabai -m window \$YABAI_WINDOW_ID --focus && yabai -m signal --remove $LABEL") && (/Applications/Alacritty.app/Contents/MacOS/alacritty msg create-window || open -na /Applications/Alacritty.app)
```

**Comment by colehpage** [comment]
Works great from my end, thanks for taking the time @koekeishiya - using a signal make a lot of sense, had a feeling that was an option in there somewhere, should have dived into the docs a bit further, but good to know moving forward. Will be curious to see if it's even an issue after the next sonoma release, will keep eyes peeled 👍

**Comment by j-c-m** [comment]
I fould my way here investigating a focus bug, slight error in the signal code, Alacritty should be lowercase. Also used pid instead of uuidgen to avoid a subshell call and prefixed it to avoid yabai thinking it was an index id not a label.

```
ctrl + cmd - a : (LABEL="pid$$" && yabai -m signal --add event="window_created" label="$LABEL" app="^alacritty$" action="yabai -m window \$YABAI_WINDOW_ID --focus && yabai -m signal --remove $LABEL") && (/opt/homebrew/bin/alacritty msg create-window || /opt/homebrew/bin/alacritty)
```
